### PR TITLE
Consensus On Epoch Boundaries

### PR DESCRIPTION
**Pull request by jordanschalm**
Adds tests and adjustments to enable the consensus committee to gracefully pass epoch boundaries, including with disjoint committees in adjacent epochs.

#### Changes
* Modifies `Voter` to not generate and process a self-vote for a block when the node has 0 voting weight at that block (addresses edge case documented [here](https://github.com/dapperlabs/flow-go/issues/4896#issuecomment-807742209)
* Add HotStuff integration test cases for epoch transitions with varying consensus committee changes
* Change `ClusterCommittee` and `ConsensusCommittee` to return `ErrInvalidSigner`, as specified by [the interface](https://github.com/onflow/flow-go/blob/f6ebd3585714bf52b7ee289c217d0c996dd762f5/consensus/hotstuff/committee.go#L31-L32)

#### To Do 

- [x] Remove `Communicator` changes 
- [x] Suppress invalid votes in `Voter`
- [x] Gracefully handle votes from invalid consensus committee members in either HotStuff or compliance engine (causes crash currently)
- [x] Committee tests/docs to test error return

### CODE REVIEW

File: consensus/hotstuff/committees/cluster_committee.go [hunk 1]
```diff
@@ -101,10 +102,10 @@ func (c *Cluster) Identity(blockID flow.Identifier, nodeID flow.Identifier) (*fl
 	// otherwise use the snapshot given by the reference block
 	identity, err := c.state.AtBlockID(payload.ReferenceBlockID).Identity(nodeID)
 	if err != nil {
-		return nil, fmt.Errorf("could not get identity for node (id=%x): %w", nodeID, err)
+		return nil, fmt.Errorf("could not get identity for node (id=%x): %w", nodeID, convertError(err))
```
Comment (AlexHentschel) [inline consensus/hotstuff/committees/cluster_committee.go:105]: With error handling, we are still struggling with failed attempts of uniform error handling patterns. I feel we should air on the opposite side for now and treat errors on a case-by case basis (no functions for uniform error handling. I am just very worried that at some point, somebody will decide to change the `convertError` method and just log the error and drop it.)

Hence, I feel we could just do the error conversion explicitly here.

File: consensus/hotstuff/committees/consensus_committee.go [hunk 2]
```diff
@@ -65,19 +66,20 @@ func NewConsensusCommittee(state protocol.State, me flow.Identifier) (*Consensus
...
 		selector,
 	))
+	return il, convertError(err)
 }
 
 func (c *Consensus) Identity(blockID flow.Identifier, nodeID flow.Identifier) (*flow.Identity, error) {
 	identity, err := c.state.AtBlockID(blockID).Identity(nodeID)
 	if err != nil {
-		return nil, fmt.Errorf("could not get identity for node ID %x: %w", nodeID, err)
+		return nil, fmt.Errorf("could not get identity for node ID %x: %w", nodeID, convertError(err))
```
Comment (AlexHentschel) [inline consensus/hotstuff/committees/consensus_committee.go:79]: same reasoning: I feel we could just do the error conversion explicitly here and remove `convertError`

File: consensus/hotstuff/committees/consensus_committee.go [hunk 3]
```diff
@@ -65,19 +66,20 @@ func NewConsensusCommittee(state protocol.State, me flow.Identifier) (*Consensus
 }
 
 func (c *Consensus) Identities(blockID flow.Identifier, selector flow.IdentityFilter) (flow.IdentityList, error) {
-	return c.state.AtBlockID(blockID).Identities(filter.And(
+	il, err := c.state.AtBlockID(blockID).Identities(filter.And(
 		filter.IsVotingConsensusCommitteeMember,
 		selector,
 	))
+	return il, convertError(err)
```
Comment (AlexHentschel) [inline consensus/hotstuff/committees/consensus_committee.go:73]: I don't think that we can reach a `protocol.IdentityNotFoundError` error here, as we only drop identities that do not pass the filter; but don't require any particular identity to be present.

File: consensus/hotstuff/eventhandler/event_handler.go [hunk 4]
```diff
@@ -376,15 +376,28 @@ func (e *EventHandler) processBlockForCurrentViewIfIsNextLeader(block *model.Blo
...
 	shouldVote := true
-	ownVote, err := e.voter.ProduceVoteIfVotable(block, curView)
-	if err != nil {
-		if !model.IsNoVoteError(err) {
-			// unknown error, exit the event loop
-			return fmt.Errorf("could not produce vote: %w", err)
-		}
-		log.Debug().Err(err).Msg("should not vote for this block")
+	_, err := e.committee.Identity(block.BlockID, e.committee.Self())
+	if errors.Is(err, model.ErrInvalidSigner) {
```
Comment (AlexHentschel) [inline consensus/hotstuff/eventhandler/event_handler.go:384]: ❗❗  
We are trying to only have logic in the event handler that is for the core HotStuff. HotStuff is already complex enough as it is. Keeping intellectual complexity as possible (by strict modularization) is the only way we can keep the code maintainable. 

This logic is _not_ part of HotStuff. I feel it should go into the `voter`

File: consensus/hotstuff/validator/validator.go [hunk 5]
```diff
@@ -137,7 +137,7 @@ func (v *Validator) ValidateVote(vote *model.Vote, block *model.Block) (*flow.Id
 		return nil, newInvalidVoteError(vote, err)
 	}
 	if err != nil {
-		return nil, fmt.Errorf("error retrieving voter Identity %x: %w", block.BlockID, err)
+		return nil, fmt.Errorf("error retrieving voter Identity (block_id=%x): %w", block.BlockID, err)
```
Comment (AlexHentschel) [inline consensus/hotstuff/validator/validator.go:140]:
```suggestion
		return nil, fmt.Errorf("error retrieving voter Identity at block %x: %w", block.BlockID, err)
```

File: engine/consensus/compliance/engine.go [hunk 6]
```diff
@@ -286,6 +286,17 @@ func (e *Engine) SendVote(blockID flow.Identifier, view uint64, sigData []byte,
...
+	staked, err := protocol.IsNodeStakedAt(e.state.AtBlockID(blockID), e.me.NodeID())
+	if err != nil {
+		// this is a fatal error and will cause HotStuff to crash
+		return fmt.Errorf("could not check my stake: %w", err)
+	}
+	if !staked {
+		e.log.Debug().Msg("skipping vote transmission for unstaked block")
+		return nil
+	}
+
```
Comment (AlexHentschel) [inline engine/consensus/compliance/engine.go:299]: If we are suppressing votes in the `Voter`, we won't need this anymore.

**Review by AlexHentschel — COMMENTED** [review]
LGTM. 

In response to the discussion in the issue [#4896](https://github.com/dapperlabs/flow-go/issues/4896#issuecomment-807742209), I have changed my mind on how to suppress votes. Sorry 😬  See also my comment here.

File: consensus/hotstuff/eventhandler/event_handler.go (hunk 4 above)
Comment (jordanschalm) [reply consensus/hotstuff/eventhandler/event_handler.go]: Moved to `Voter`

**Comment by jordanschalm** [comment]
@AlexHentschel @zhangchiqing This is ready for review

File: consensus/hotstuff/voter/voter.go [hunk 7]
```diff
@@ -45,6 +55,16 @@ func (v *Voter) ProduceVoteIfVotable(block *model.Block, curView uint64) (*model
 		return nil, model.NoVoteError{Msg: "not above the last voted view"}
 	}
 
+	// Do not produce a vote for blocks where we are not a valid committee
+	// member. These votes can't be used to produce valid QCs.
```
Comment (zhangchiqing) [inline consensus/hotstuff/voter/voter.go:59]:
```suggestion
	// member. HotStuff will ask for a vote when cross the epoch boundary. 
	// These votes can't be used to produce valid QCs.
```

**Review by zhangchiqing — APPROVED** [review]
Looks good 👍

File: consensus/hotstuff/voter/voter.go (hunk 7 above)
Comment (jordanschalm) [reply consensus/hotstuff/voter/voter.go:59]: I don't quite understand this. HotStuff asks `Voter` for a vote for all blocks, both before and after the epoch boundary.

Comment (zhangchiqing) [reply consensus/hotstuff/voter/voter.go:59]: good point. updated.

File: consensus/hotstuff/committees/consensus_committee_test.go [hunk 8]
```diff
@@ -16,6 +17,57 @@ import (
...
+	t.Run("non-existent identity should return ErrInvalidSigner", func(t *testing.T) {
+		_, err := com.Identity(blockID, fakeID)
+		require.True(t, errors.Is(model.ErrInvalidSigner, err))
+	})
+
+	t.Run("existent but non-committee-member identity should return ErrInvalidSigner", func(t *testing.T) {
+		_, err := com.Identity(blockID, realNonCommitteeIdentity.NodeID)
+		require.True(t, errors.Is(model.ErrInvalidSigner, err))
+	})
+
```
Comment (AlexHentschel) [inline consensus/hotstuff/committees/consensus_committee_test.go:63]: Thank you for extending the tests. While we check that:
* nodes with zero staked are filtered out

I think there are more edge-cases, which we could check:
* nodes with role _other_ than `RoleConsensus` are filtered out 
* ejected consensus nodes are filtered out

File: consensus/integration/nodes_test.go [hunk 9]
```diff
@@ -11,6 +10,7 @@ import (
 	"github.com/stretchr/testify/mock"
 	"github.com/stretchr/testify/require"
 
+	"github.com/onflow/flow-go/model/flow/filter"
 	consensusMempools "github.com/onflow/flow-go/module/mempool/consensus"
 
 	"github.com/onflow/flow-go/consensus"
```
Comment (AlexHentschel) [inline consensus/integration/nodes_test.go:16]: unordered inputs

**Review by AlexHentschel — APPROVED** [review]
Very nice clean code. Thank you.

File: consensus/hotstuff/committees/consensus_committee_test.go (hunk 8 above)
Comment (jordanschalm) [reply consensus/hotstuff/committees/consensus_committee_test.go:79]: Added cases in https://github.com/onflow/flow-go/pull/565/commits/dabbbfeda1d65102dcb288af356c1933168b2e99

File: consensus/integration/nodes_test.go (hunk 9 above)
Comment (jordanschalm) [reply consensus/integration/nodes_test.go]: https://github.com/onflow/flow-go/pull/565/commits/506e7e2f74f4a8ab5707b0c7d62c5e31f0c5d49e

**Pull request merged**